File: wyrm/lexer.py
import re
from typing import List, Tuple, Optional, Union

# Token types
TT_INT = 'INT'
TT_FLOAT = 'FLOAT'
TT_STRING = 'STRING'
TT_IDENTIFIER = 'IDENTIFIER'
TT_KEYWORD = 'KEYWORD'
TT_OPERATOR = 'OPERATOR'
TT_DELIMITER = 'DELIMITER'
TT_COMMENT = 'COMMENT'
TT_NEWLINE = 'NEWLINE'
TT_EOF = 'EOF'

# Keywords
KEYWORDS = {
    'fn', 'true', 'false', 'null',
    'if', 'elif', 'else',
    'repeat', 'til',
    'break', 'continue',
    'print', 'input',
    'int', 'float', 'str',
    'len', 'append', 'pop',
    'return',
    'var', 'dec', 'use'
}

# Operators - characters that can start an operator
OPERATORS = {'+', '-', '*', '/', '%', '=', '!', '<', '>', '&', '|'}

# Complete operators (for checking what we've built)
OPERATOR_KEYS = {
    '+', '-', '*', '/', '//', '%',
    '=', '+=', '-=', '*=', '/=', '%=',
    '==', '!=', '<', '>', '<=', '>=',
    '&', '&&', '|', '||',
    '!',
    'and', 'or', 'not'
}

# Delimiters
DELIMITERS = {
    '(', ')', '{', '}', '[', ']',
    ',', ':', ';', '.'
}

class Token:
    def __init__(self, type: str, value: Union[str, int, float], line: int, column: int):
        self.type = type
        self.value = value
        self.line = line
        self.column = column

    def __repr__(self):
        return f"Token({self.type}, {repr(self.value)}, line={self.line}, col={self.column})"

class Lexer:
    def __init__(self, text: str):
        self.text = text
        self.pos = 0
        self.line = 1
        self.column = 1
        self.current_char = self.text[0] if self.text else None

    def advance(self):
        """Advance to the next character."""
        if self.current_char == '\n':
            self.line += 1
            self.column = 1
        elif self.current_char is not None:
            self.column += 1

        self.pos += 1
        if self.pos >= len(self.text):
            self.current_char = None
        else:
            self.current_char = self.text[self.pos]

    def peek(self, n: int = 1) -> Optional[str]:
        """Look ahead n characters without consuming."""
        peek_pos = self.pos + n
        if peek_pos >= len(self.text):
            return None
        return self.text[peek_pos]

    def skip_whitespace(self):
        """Skip whitespace characters."""
        while self.current_char is not None and self.current_char.isspace():
            self.advance()

    def skip_comment(self):
        """Skip single-line comment."""
        while self.current_char is not None and self.current_char != '\n':
            self.advance()

    def read_number(self) -> Token:
        """Read a number (int or float)."""
        result = ''
        dot_count = 0

        while self.current_char is not None and (self.current_char.isdigit() or self.current_char == '.'):
            if self.current_char == '.':
                dot_count += 1
                if dot_count > 1:
                    break  # Only one decimal point allowed
            result += self.current_char
            self.advance()

        if dot_count == 1:
            return Token(TT_FLOAT, float(result), self.line, self.column - len(result))
        else:
            return Token(TT_INT, int(result), self.line, self.column - len(result))

    def read_string(self) -> Token:
        """Read a string literal."""
        quote_char = self.current_char  # Either " or '
        self.advance()  # Skip opening quote

        result = ''
        while self.current_char is not None and self.current_char != quote_char:
            if self.current_char == '\\':  # Handle escape sequences
                self.advance()
                if self.current_char == 'n':
                    result += '\n'
                elif self.current_char == 't':
                    result += '\t'
                elif self.current_char == 'r':
                    result += '\r'
                elif self.current_char == '\\':
                    result += '\\'
                elif self.current_char == quote_char:
                    result += quote_char
                else:
                    result += self.current_char  # Unknown escape, keep as-is
            else:
                result += self.current_char
            self.advance()

        if self.current_char == quote_char:
            self.advance()  # Skip closing quote
        return Token(TT_STRING, result, self.line, self.column - len(result) - 1)

    def read_identifier(self) -> Token:
        """Read an identifier or keyword."""
        result = ''
        while self.current_char is not None and (self.current_char.isalnum() or self.current_char == '_'):
            result += self.current_char
            self.advance()

        if result in KEYWORDS:
            return Token(TT_KEYWORD, result, self.line, self.column - len(result))
        else:
            return Token(TT_IDENTIFIER, result, self.line, self.column - len(result))

    def tokenize(self) -> List[Token]:
        """Convert input text into a list of tokens."""
        tokens = []

        while self.current_char is not None:
            if self.current_char.isspace():
                if self.current_char == '\n':
                    tokens.append(Token(TT_NEWLINE, '\n', self.line, self.column))
                self.advance()

            elif self.current_char.isdigit():
                tokens.append(self.read_number())

            elif self.current_char in ('"', "'"):
                tokens.append(self.read_string())

            elif self.current_char.isalpha() or self.current_char == '_':
                tokens.append(self.read_identifier())

            elif self.current_char == '/' and self.peek() == '/':
                # Single-line comment
                self.skip_comment()
                # Don't add comment token for now (can be added if needed)

            elif self.current_char in OPERATORS:
                # Find the longest matching operator starting at current position
                op = ""
                # Check progressively longer substrings
                for i in range(1, min(10, len(self.text) - self.pos + 1)):  # up to 10 chars or end of string
                    substring = self.text[self.pos:self.pos + i]
                    if substring in OPERATOR_KEYS:
                        op = substring  # keep the longest match found
                    else:
                        # If this substring isn't an operator, longer ones won't be either
                        break

                if op:  # we found a matching operator
                    # Record starting position
                    start_line = self.line
                    start_column = self.column
                    # Consume the operator characters
                    for _ in range(len(op)):
                        self.advance()
                    tokens.append(Token(TT_OPERATOR, op, start_line, start_column))
                else:
                    # Fallback: treat as single character (shouldn't happen)
                    start_line = self.line
                    start_column = self.column
                    self.advance()
                    tokens.append(Token(TT_OPERATOR, self.current_char, start_line, start_column))

            elif self.current_char in DELIMITERS:
                tokens.append(Token(TT_DELIMITER, self.current_char, self.line, self.column))
                self.advance()

            else:
                # Unknown character
                raise SyntaxError(f"Unknown character: {self.current_char} at line {self.line}, column {self.column}")

        tokens.append(Token(TT_EOF, None, self.line, self.column))
        return tokens

File: compiler/lexer/lexer.cpp
#include "lexer.hpp"
#include <stdexcept>
#include <cctype>
#include <algorithm>

namespace wyrm {

const std::unordered_set<std::string_view> Lexer::KEYWORDS = {
    "fn", "true", "false", "null",
    "if", "elif", "else",
    "repeat", "do", "til",
    "break", "continue",
    "print", "input",
    "int", "float", "str",
    "len", "append", "pop",
    "return",
    "var", "dec", "use",
    "unsafe", "owned", "arena"
};

const std::unordered_set<char> Lexer::OPERATOR_CHARS = {
    '+', '-', '*', '/', '%', '=', '!', '<', '>', '&', '|'
};

const std::unordered_set<std::string_view> Lexer::OPERATOR_KEYS = {
    "+", "-", "*", "/", "//", "%",
    "=", "+=", "-=", "*=", "/=", "%=",
    "==", "!=", "<", ">", "<=", ">=",
    "&", "&&", "|", "||",
    "!",
    "and", "or", "not"
};

const std::unordered_set<char> Lexer::DELIMITERS = {
    '(', ')', '{', '}', '[', ']',
    ',', ':', ';', '.'
};

Lexer::Lexer(std::string_view source)
    : source_(source), pos_(0), line_(1), column_(1) {}

char Lexer::current_char() const {
    if (pos_ >= source_.size()) {
        return '\0';
    }
    return source_[pos_];
}

char Lexer::peek(size_t offset) const {
    if (pos_ + offset >= source_.size()) {
        return '\0';
    }
    return source_[pos_ + offset];
}

void Lexer::advance() {
    char curr = current_char();
    if (curr == '\0') {
        return;
    }
    if (curr == '\n') {
        line_++;
        column_ = 1;
    } else {
        column_++;
    }
    pos_++;
}

void Lexer::skip_whitespace() {
    while (true) {
        char curr = current_char();
        if (curr != '\0' && curr != '\n' && std::isspace(static_cast<unsigned char>(curr))) {
            advance();
        } else {
            break;
        }
    }
}

void Lexer::skip_comment() {
    while (true) {
        char curr = current_char();
        if (curr != '\0' && curr != '\n') {
            advance();
        } else {
            break;
        }
    }
}

Token Lexer::read_number() {
    size_t start_col = column_;
    std::string result;
    size_t dot_count = 0;

    while (true) {
        char curr = current_char();
        if (curr == '\0') {
            break;
        }
        if (std::isdigit(static_cast<unsigned char>(curr))) {
            result += curr;
            advance();
        } else if (curr == '.') {
            // Check if next character is a digit to differentiate member access like 'obj.method' from float '.5'
            char next = peek(1);
            if (std::isdigit(static_cast<unsigned char>(next)) && dot_count == 0) {
                dot_count++;
                result += curr;
                advance();
            } else {
                break; // Delimiter '.' or invalid dot sequence
            }
        } else {
            break;
        }
    }

    if (dot_count == 1) {
        return Token(TokenType::FLOAT, std::move(result), line_, start_col);
    } else {
        return Token(TokenType::INT, std::move(result), line_, start_col);
    }
}

Token Lexer::read_string() {
    size_t start_col = column_;
    char quote_char = current_char(); // Either " or '
    advance(); // Skip quote

    std::string result;
    while (true) {
        char curr = current_char();
        if (curr == '\0') {
            throw std::runtime_error("SyntaxError: Unterminated string literal at line " + std::to_string(line_) + ", col " + std::to_string(start_col));
        }
        if (curr == quote_char) {
            advance(); // Skip closing quote
            break;
        }
        if (curr == '\\') {
            advance();
            char escaped = current_char();
            if (escaped == '\0') {
                throw std::runtime_error("SyntaxError: Unterminated string escape sequence at line " + std::to_string(line_) + ", col " + std::to_string(column_));
            }
            if (escaped == 'n') {
                result += '\n';
            } else if (escaped == 't') {
                result += '\t';
            } else if (escaped == 'r') {
                result += '\r';
            } else if (escaped == '\\') {
                result += '\\';
            } else if (escaped == quote_char) {
                result += quote_char;
            } else {
                result += escaped; // Unknown escape sequence, keep as is
            }
        } else {
            result += curr;
        }
        advance();
    }

    return Token(TokenType::STRING, std::move(result), line_, start_col);
}

Token Lexer::read_identifier() {
    size_t start_col = column_;
    std::string result;

    while (true) {
        char curr = current_char();
        if (curr == '\0') {
            break;
        }
        if (std::isalnum(static_cast<unsigned char>(curr)) || curr == '_') {
            result += curr;
            advance();
        } else {
            break;
        }
    }

    if (KEYWORDS.contains(result)) {
        return Token(TokenType::KEYWORD, std::move(result), line_, start_col);
    } else {
        return Token(TokenType::IDENTIFIER, std::move(result), line_, start_col);
    }
}

Token Lexer::read_operator() {
    size_t start_col = column_;
    std::string op;

    // Greedy search for longest valid operator starting at current position
    for (size_t i = 1; i <= 10; ++i) {
        if (pos_ + i > source_.size()) {
            break;
        }
        std::string_view sub = source_.substr(pos_, i);
        if (OPERATOR_KEYS.contains(sub)) {
            op = std::string(sub);
        } else {
            break;
        }
    }

    if (!op.empty()) {
        for (size_t k = 0; k < op.size(); ++k) {
            advance();
        }
        return Token(TokenType::OPERATOR, std::move(op), line_, start_col);
    }

    // Fallback if no operator matches but character is in OPERATOR_CHARS
    std::string fallback(1, current_char());
    advance();
    return Token(TokenType::OPERATOR, std::move(fallback), line_, start_col);
}

std::vector<Token> Lexer::tokenize() {
    std::vector<Token> tokens;

    while (true) {
        char curr = current_char();
        if (curr == '\0') {
            break;
        }

        if (curr != '\n' && std::isspace(static_cast<unsigned char>(curr))) {
            skip_whitespace();
        } else if (curr == '\n') {
            tokens.emplace_back(TokenType::NEWLINE, "\n", line_, column_);
            advance();
        } else if (std::isdigit(static_cast<unsigned char>(curr))) {
            tokens.push_back(read_number());
        } else if (curr == '"' || curr == '\'') {
            tokens.push_back(read_string());
        } else if (std::isalpha(static_cast<unsigned char>(curr)) || curr == '_') {
            // Read identifier, check if it's a word operator 'and', 'or', 'not' first
            Token tok = read_identifier();
            if (tok.type == TokenType::KEYWORD && (tok.value == "and" || tok.value == "or" || tok.value == "not")) {
                tok.type = TokenType::OPERATOR;
            }
            tokens.push_back(std::move(tok));
        } else if (curr == '/' && peek(1) == '/') {
            // Single-line comment: check prefix rules
            if (!tokens.empty() && tokens.back().type != TokenType::NEWLINE) {
                throw std::runtime_error("SyntaxError: Unexpected comment start '//' at line " + std::to_string(line_) + ", col " + std::to_string(column_));
            }
            skip_comment();
        } else if (OPERATOR_CHARS.contains(curr)) {
            tokens.push_back(read_operator());
        } else if (DELIMITERS.contains(curr)) {
            std::string delim(1, curr);
            tokens.emplace_back(TokenType::DELIMITER, std::move(delim), line_, column_);
            advance();
        } else {
            throw std::runtime_error(std::string("SyntaxError: Unknown character: ") + curr + " at line " + std::to_string(line_) + ", col " + std::to_string(column_));
        }
    }

    tokens.emplace_back(TokenType::END_OF_FILE, "", line_, column_);
    return tokens;
}

} // namespace wyrm
